### PR TITLE
Initial objective-c extern implementation

### PR DESCRIPTION
**Pull request by waneck**
This is just a quick sketch of a feature discussed with @hughsando at WWX. Ultimately, this should allow to interoperate with Objective-C code directly from Haxe, by defining special extern classes (in this PR, by adding the `@:objc` meta).

At the moment, only function calls are implemented. This PR should bootstrap some of the changes that HaxeFoundation/hxcpp might need in order to support correct boxing/unboxing of Objective-C types.

@hughsando, please let me know if you agree with the approach, and if there's anything you think that needs to be changed.
I haven't implemented the plain C calls (also since I know there will be a linking problem there, since the included files aren't extern "C"), but I'm thinking about using this `@:plain` metadata.

I have the following roadmap in mind:
- [x] Initial implementation and function calling
- [x] NSObject boxing
- [x] test interfaces ("protocols" in objective-c)
- [ ] basic types mapping (e.g. enums, NSNumber, NSArray, etc)
- [ ] add/test C externs as well (test extern C `@:include`)
- [ ] add framework referencing through metadata (+ check with DCE on)
- [ ] automatically add some defines to select between implementations (e.g. compile objective-c code path only when compiling to mac/ios) - cross-compilation aware
- [ ] Haxe functions -> Objc blocks
- [ ] Reference couting and GC interop
- [ ] [try/catch](https://developer.apple.com/library/ios/documentation/Cocoa/Conceptual/Exceptions/Articles/Exceptions64Bit.html#//apple_ref/doc/uid/TP40009044-SW3)
- [ ] extend objective-c classes
- [ ] implement objective-c protocols
- [ ] minor details: haxe `Class<T>` and objective-c `Class` interop, and more testing
- [ ] start to meddle with C++ extern support
- [ ] add `@:objcpp` to indicate headers that should be included without extern C

I'm thinking about adding them by parts so we can discuss each implementation in a way that is graspable and readable, and not a all-in implementation

### CODE REVIEW

**Comment by waneck** [comment]
I've just included another commit. Starting from this one (commit 4e906a1), all subsequent commits will fail until HaxeFoundation/hxcpp#245 is merged.

The following commit message describes the changes:
```
    Add initial boxing tests, and guard changes on `-D objc` define
    
    The tests contained herein will only pass once HaxeFoundation/hxcpp#245
    is accepted. It adds basic tests for the boxing and unboxing operations,
    and it adds some extra code to handle `null`, and a few operators
    correctly.
    
    Also, this makes sure that `@:objc` code is only run when `-D objc` is
    defined - which should make any changes safe against regressions. I've
    also made sure any added code is labeled `objc` properly.
```

File: gencpp.ml [hunk 1]
```diff
@@ -1801,6 +1823,20 @@ and gen_expression ctx retval expression =
...
 
+   (* this will add a cast if boxing / unboxing an objective-c type *)
+   let check_objc_unbox expression to_type =
+     if is_objc_type to_type && not (is_objc_type expression.etype) then
+        { expression with eexpr = TCast(expression,None); etype = to_type }
+     else
+        expression
+   in
+   let check_objc_box expression to_type =
+     if is_objc_type expression.etype && not (is_objc_type to_type) then
```
Comment (waneck) [inline gencpp.ml]: This `not (is_objc_type to_type)` should probably be replaced by `is_dynamic_in_cpp`. I'll add more tests and make the change if needed

**Comment by hughsando** [comment]
I'm pinging this request to see what travis thinks now I've fixed hxcpp

**Comment by Simn** [comment]
Were you supposed to merge this already?

**Comment by hughsando** [comment]
My understanding is that this code will not be reached unless the special
meta is used in the source, so it should not break anything.  That being
the case, I would prefer to merge early to avoid conflicts and also I might
have some ideas on certain things.

However, this can be reverted if you think its best without any stress for
me.

**Comment by Simn** [comment]
This was more a question for @waneck I guess. I don't mind either way.

**Comment by waneck** [comment]
Sorry to take some time to answer
First of all, @hughsando, thanks for accepting this PR. I've had to work on a couple of things but I'm back working on objective-c support now.
There are indeed still things left to be done now - and I plan to keep working on them and keep sending PRs as I go :)

As for merging them now, I'm of course ok with that. The problem with long-standing PRs is that the chance of them becoming outdated and need manual merges get higher as the time passes. And all changes are very self-contained within specific flags (which is something I plan to do throughout this feature)

**Comment by jeremyfa** [comment]
You guys are doing fantastic work! Can't wait to see how this feature will evolve :+1:

**Pull request merged**